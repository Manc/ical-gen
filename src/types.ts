export interface ICalRepeatingOptions {
	freq: ICalEventRepeatingFreq;
	count?: number;
	interval?: number;
	until?: Date;
	byDay?: ICalWeekday[] | ICalWeekday;
	byMonth?: number[] | number;
	byMonthDay?: number[] | number;
	bySetPos?: number;
	exclude?: Date[] | Date;
	startOfWeek?: ICalWeekday;
}

export interface ICalGeo {
	lon: number;
	lat: number;
}

export interface ICalLocation {
	title: string;
	address?: string;
	radius?: number;
	geo?: ICalGeo;
}

export interface ICalOrganizer {
	name: string;
	email?: string;
	mailto?: string;
}

export interface ICalDescription {
	plain: string;
	html?: string;
}

export type ICalEventRepeatingFreq =
	| 'SECONDLY'
	| 'MINUTELY'
	| 'HOURLY'
	| 'DAILY'
	| 'WEEKLY'
	| 'MONTHLY'
	| 'YEARLY'
;

export type ICalWeekday =
	| 'MO'
	| 'TU'
	| 'WE'
	| 'TH'
	| 'FR'
	| 'SA'
	| 'SU'
;
