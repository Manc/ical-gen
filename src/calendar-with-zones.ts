import type ICalComponent from './component';
import ICalCalendar, { ICalCalendarOptions } from './calendar';
import type ICalTimeZone from './timezone';

/**
 * An async function that, given a time zone string (e.g. `America/Los_Angeles`)
 * either resolves with a `ICalTimeZone` component (`VTIMEZONE`) or rejects the
 * Promise.
 */
export type ICalVTimezoneGenerator = (timezone: string) => Promise<ICalTimeZone>;

export type ICalCalendarWithZonesOptions = ICalCalendarOptions & {
	/**
	 * An async function that, given a time zone string (e.g. `America/Los_Angeles`)
	 * either resolves with a `ICalTimeZone` component (`VTIMEZONE`) or rejects the
	 * Promise.
	 */
	vTimeZoneGenerator: ICalVTimezoneGenerator;
};

export default class ICalCalendarWithZones extends ICalCalendar {
	public vTimeZoneGenerator: ICalVTimezoneGenerator;

	constructor(options: ICalCalendarWithZonesOptions) {
		super(options);
		this.vTimeZoneGenerator = options.vTimeZoneGenerator;
	}

	/**
	 * This implementation of `getChildrenForRender()` resolves all time zones
	 * to time zone components and dynamically adds them in front of all other
	 * other children.
	 */
	public getChildrenForRender = async (): Promise<Readonly<Array<ICalComponent | string>>> => {
		// Get all time zones, but ignore UTC.
		const timeZones = this.getTimeZones().filter(tz => tz !== 'UTC');

		// Try to generate a `VTIMEZONE` string for each time zone.
		const tzComponents: Array<ICalTimeZone> = (
			await Promise.all(
				timeZones.map(tz =>
					this.vTimeZoneGenerator(tz).catch(() => undefined)
				)
			)
		).filter(component => !!component) as ICalTimeZone[];

		return [
			...tzComponents, // Insert `VTIMEZONE`s first
			...this.children, // Then the rest like `VENENT` etc.
		];
	};
}
