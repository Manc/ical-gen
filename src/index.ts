// import ICalCalendar from './calendar';
// import type { CalendarOptions } from './calendar';
// import type ICalEvent from './event';

// export {
// 	default as ICalAlarm,
// 	ICalAlarmData,
// 	ICalAlarmType,
// 	ICalAlarmTypeValue,
// 	ICalAlarmJSONData,
// 	ICalAttachment
// } from './alarm';

export {
	default as ICalComponent,
	ICalComponentOptions,
} from './component';

export {
	default as ICalAttendee,
	ICalAttendeeData,
	ICalAttendeeType,
	ICalAttendeeRole,
	ICalAttendeeStatus,
	ICalAttendeeJSONData
} from './attendee';

export {
	default as ICalCalendar,
	ICalCalendarOptions,
	ICalCalendarProdId,
	ICalCalendarMethod,
} from './calendar';

export {
	default as ICalCalendarWithZones,
	ICalCalendarWithZonesOptions,
	ICalVTimezoneGenerator,
} from './calendar-with-zones';

export {
	default as ICalTimeZone,
	ICalTimeZoneOptions,
} from './timezone';

export {
	default as ICalCategory,
	ICalCategoryData
} from './category';

export {
	default as ICalEvent,
	ICalEventOptions,
	ICalEventStatus,
	ICalEventBusyStatus,
	ICalEventTransparency,
	// ICalEventJSONData
} from './event';

export {
	// ICalInternalDateTimeValue as ICalDateTimeValue,
	ICalRepeatingOptions,
	ICalLocation,
	ICalGeo,
	ICalOrganizer,
	ICalDescription,
	ICalEventRepeatingFreq,
	ICalWeekday,
	// ICalTimezone
} from './types';

export {
	// formatDateOld as formatDate,
	// formatDateTZ,
	formatDateUTC,
	formatDateProp,
	escapeICalValue,
	foldLine,
	filterAndJoinICalLines,
} from './utils';
