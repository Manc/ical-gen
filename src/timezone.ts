import ICalComponent from './component';

export type ICalTimeZoneOptions = {
	fromString: string;
} | {
	fromLines: string[];
}

export default class ICalTimeZone extends ICalComponent {
	public tagName: string = 'VTIMEZONE';
	private lines: string[];

	constructor(options: ICalTimeZoneOptions) {
		super();
		if ('fromLines' in options) {
			this.lines = options.fromLines;
		} else {
			this.lines = options.fromString.split('\n').map(l => l.trimEnd());
		}
	}

	public getTimeZones = (): string[] => [];

	protected renderToLines = async (): Promise<string[]> => this.lines;
}
