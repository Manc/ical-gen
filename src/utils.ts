import { DateTime } from 'luxon';
import { ICalOrganizer } from './types';

const CRLF = '\r\n';

export type ICalZonedDate = DateTime | {
	/**
	 * Date as absolute moment in time. Note that the JS-native Date does
	 * not contain any time zone information.
	 */
	date: Date;

	/**
	 * Time zone. If undefined, UTC will be assumed.
	 * 
	 * @example America/Los_Angeles
	 */
	zone?: string;
};

export function getTimeZoneStr(zonedDate: ICalZonedDate): string | undefined {
	return zonedDate instanceof DateTime
		? zonedDate.zoneName
		: zonedDate.zone;
}

function zondedDateToDateTime(zonedDate: ICalZonedDate): DateTime {
	return zonedDate instanceof DateTime
		? zonedDate
		: DateTime.fromJSDate(zonedDate.date, { zone: zonedDate.zone || 'UTC' });
}

export function formatZonedDate(zonedDate: ICalZonedDate, dateOnly = false): string {
	const dt = zondedDateToDateTime(zonedDate);
	return dt.toFormat(dateOnly
		? 'yyyyLLdd'
		: 'yyyyLLdd\'T\'HHmmss'
	);
}

export function formatDateUTC(date: Date, dateOnly = false): string {
	return formatZonedDate({ date }, dateOnly) + 'Z';
}

// export function formatZonedDateUTC(zonedDate: ICalZonedDate, dateOnly = false): string {
// 	const dt = zondedDateToDateTime(zonedDate);
// 	return dt.toFormat(dateOnly
// 		? `yyyyLLdd`
// 		: `yyyyLLdd'T'HHmmss`
// 	);
// }

// export function formatDateUTC(d: Date, dateOnly = false): string {
// 	return (
// 		d.getUTCFullYear().toString() +
// 		(d.getUTCMonth() + 1).toString().padStart(2, '0') +
// 		d.getUTCDate().toString().padStart(2, '0') +
// 		dateOnly
// 			? ''
// 			: (
// 				d.getUTCHours().toString().padStart(2, '0') +
// 				d.getUTCMinutes().toString().padStart(2, '0') +
// 				d.getUTCSeconds().toString().padStart(2, '0')
// 			)
// 	);
// 	// return (
// 	// 	d.getFullYear().toString() +
// 	// 	(d.getMonth() + 1).toString().padStart(2, '0') +
// 	// 	d.getDate().toString().padStart(2, '0') +
// 	// 	dateOnly
// 	// 		? ''
// 	// 		: (
// 	// 			d.getHours().toString().padStart(2, '0') +
// 	// 			d.getMinutes().toString().padStart(2, '0') +
// 	// 			d.getSeconds().toString().padStart(2, '0')
// 	// 		)
// 	// );
// }

// /**
//  * Converts a valid date/time object supported by this library to a string.
//  * For information about this format, see RFC 5545, section 3.3.5
//  * https://tools.ietf.org/html/rfc5545#section-3.3.5
//  * 
//  * @deprecated Switch to `formatTZDate()`.
//  */
// export function formatDateTZ(hasTimeZone: boolean, property: string, date: Date, eventData?: {floating?: boolean | null, timezone?: string | null}): string {
// 	let tzParam = '';
// 	let floating = eventData?.floating || false;

// 	if (eventData?.timezone) {
// 		tzParam = ';TZID=' + eventData.timezone;

// 		// This isn't a 'floating' event because it has a timezone;
// 		// but we use it to omit the 'Z' UTC specifier in formatDate()
// 		floating = true;
// 	}

// 	return property + tzParam + ':' + formatDateOld(hasTimeZone, date, false, floating);
// }

export function formatDateProp(propertyName: string, zonedDate: ICalZonedDate, floating: boolean, allDay: boolean): string {
	if (floating || allDay) {
		return `${propertyName}${allDay ? ';VALUE=DATE' : ''}:${formatZonedDate(zonedDate, allDay)}`;
	} else {
		const dt = zondedDateToDateTime(zonedDate);
		const zone = dt.zone.name;
		return zone === 'UTC'
			? `${propertyName}:${formatZonedDate(zonedDate, false)}Z`
			: `${propertyName};TZID=${zone}:${formatZonedDate(zonedDate, false)}`;
	}
}

/**
 * Escape special characters in a string intended to be used as value in an
 * iCalendar file.
 */
export function escapeICalValue(str: string): string {
	return str.replace(/[\\;,"]/g, function (match) {
		return '\\' + match;
	}).replace(/(?:\r\n|\r|\n)/g, '\\n');
}

/**
 * Cleans up an array of unfiltered lines intendet to be used as lines of an
 * iCalendar file. This function should usually be used after generating the
 * lines and before calling `calLinesToString()`.
 */
export function cleanLines(lines: Array<string | undefined>): string[] {
	return lines
		.map(l => l?.trimEnd()) // Trim end of lines
		.filter(l => !!l) as string[]; // Remove empty lines including `undefined` values
}

export function foldLine(line: string): string {
	const str = line as string;
	let result = '';
	let c = 0;
	for (let i = 0; i < str.length; i++) {
		let ch = str.charAt(i);

		// surrogate pair, see https://mathiasbynens.be/notes/javascript-encoding#surrogate-pairs
		if (ch >= '\ud800' && ch <= '\udbff') {
			ch += str.charAt(++i);
		}

		const charsize = Buffer.from(ch, 'utf-8').length;
		c += charsize;
		if (c > 74) {
			result += `${CRLF} `;
			c = charsize;
		}

		result += ch;
	}
	return result;
}

/**
 * Converts iCalendar lines to one string with proper lines breaks and line-wrapping.
 */
export function filterAndJoinICalLines(lines: Array<string | undefined>): string {
	return filterEmpty(lines)
		// .map(foldLine)
		.join(CRLF);
}

export function addOrGetCustomAttributes (data: {x: [string, string][]}, keyOrArray: ({key: string, value: string})[] | [string, string][] | Record<string, string>): void;
export function addOrGetCustomAttributes (data: {x: [string, string][]}, keyOrArray: string, value: string): void;
export function addOrGetCustomAttributes (data: {x: [string, string][]}): ({key: string, value: string})[];
export function addOrGetCustomAttributes (data: {x: [string, string][]}, keyOrArray?: ({key: string, value: string})[] | [string, string][] | Record<string, string> | string  | undefined, value?: string | undefined): void | ({key: string, value: string})[] {
	if (Array.isArray(keyOrArray)) {
		data.x = keyOrArray.map((o: {key: string, value: string} | [string, string]) => {
			if(Array.isArray(o)) {
				return o;
			}
			if (typeof o.key !== 'string' || typeof o.value !== 'string') {
				throw new Error('Either key or value is not a string!');
			}
			if (o.key.substr(0, 2) !== 'X-') {
				throw new Error('Key has to start with `X-`!');
			}

			return [o.key, o.value] as [string, string];
		});
	}
	else if (typeof keyOrArray === 'object') {
		data.x = Object.entries(keyOrArray).map(([key, value]) => {
			if (typeof key !== 'string' || typeof value !== 'string') {
				throw new Error('Either key or value is not a string!');
			}
			if (key.substr(0, 2) !== 'X-') {
				throw new Error('Key has to start with `X-`!');
			}

			return [key, value];
		});
	}
	else if (typeof keyOrArray === 'string' && typeof value === 'string') {
		if (keyOrArray.substr(0, 2) !== 'X-') {
			throw new Error('Key has to start with `X-`!');
		}

		data.x.push([keyOrArray, value]);
	}
	else {
		return data.x.map(a => ({
			key: a[0],
			value: a[1]
		}));
	}
}

// export function attributesToLines(attributes: Record<string, string>): string[] {
// 	return Object.entries(attributes)
// 		.map(([key, value]) => `${key.toUpperCase()}:${escapeICalValue(value)}`);
// }

/**
 * Check the given string or ICalOrganizer. Parses
 * the string for name and email address if possible.
 *
 * @param attribute Attribute name for error messages
 * @param value Value to parse name/email from
 */
export function checkNameAndMail(attribute: string, value: string | ICalOrganizer): ICalOrganizer {
	let result: ICalOrganizer | null = null;

	if (typeof value === 'string') {
		const match = value.match(/^(.+) ?<([^>]+)>$/);
		if (match) {
			result = {
				name: match[1].trim(),
				email: match[2].trim()
			};
		}
		else if(value.includes('@')) {
			result = {
				name: value.trim(),
				email: value.trim()
			};
		}
	}
	else if (typeof value === 'object') {
		result = {
			name: value.name,
			email: value.email,
			mailto: value.mailto
		};
	}

	if (!result && typeof value === 'string') {
		throw new Error(
			'`' + attribute + '` isn\'t formated correctly. See https://sebbo2002.github.io/ical-generator/develop/'+
			'reference/interfaces/icalorganizer.html'
		);
	}
	else if (!result) {
		throw new Error(
			'`' + attribute + '` needs to be a valid formed string or an object. See https://sebbo2002.github.io/'+
			'ical-generator/develop/reference/interfaces/icalorganizer.html'
		);
	}

	if (!result.name) {
		throw new Error('`' + attribute + '.name` is empty!');
	}

	return result;
}

/**
 * Checks if the given string `value` is a
 * valid one for the type `type`
 */
export function checkEnum(type: Record<string, string>, value: unknown): unknown {
	const allowedValues = Object.values(type);
	const valueStr = String(value).toUpperCase();

	if (!valueStr || !allowedValues.includes(valueStr)) {
		throw new Error(`Input must be one of the following: ${allowedValues.join(', ')}`);
	}

	return valueStr;
}

// /**
//  * Checks if the given input is a valid timestamp or `Date` and
//  * returns the internal representation in Epoch seconds.
//  */
// export function parseInputDate(value: ICalDateTime): Date {
// 	if (value instanceof Date) {
// 		return parseInputDate(value.getTime());
// 	}


// 	if (typeof value === 'number') {
// 		if (value > 0) {
// 			return value < 1e14
// 				? value
// 				: Math.floor(value / 1e3); // Interpret value as milliseconds and convert to seconds.
// 		}
// 	} else if (value instanceof Date) {
// 		return parseInputDate(value.getTime());
// 	}

// 	throw new Error('Invalid Date');
// }

// export function toDate(value: ICalInternalDateTimeValue): Date {
// 	if (typeof value === 'string' || value instanceof Date) {
// 		return new Date(value);
// 	}

// 	// @ts-ignore
// 	if(isLuxonDate(value)) {
// 		return value.toJSDate();
// 	}

// 	return value.toDate();
// }

// export function isMoment(value: ICalDateTimeValue): value is Moment {

// 	// @ts-ignore
// 	return value != null && value._isAMomentObject != null;
// }
// export function isMomentTZ(value: ICalDateTimeValue): value is MomentTZ {
// 	return isMoment(value) && typeof value.tz === 'function';
// }
// export function isDayjs(value: ICalDateTimeValue): value is Dayjs {
// 	return typeof value === 'object' &&
// 		value !== null &&
// 		!(value instanceof Date) &&
// 		!isMoment(value) &&
// 		!isLuxonDate(value);
// }
// export function isLuxonDate(value: ICalDateTimeValue): value is LuxonDateTime {
// 	return typeof value === 'object' && value !== null && typeof (value as LuxonDateTime).toJSDate === 'function';
// }

// export function isMomentDuration(value: unknown): value is Duration {

// 	// @ts-ignore
// 	return value !== null && typeof value === 'object' && typeof value.asSeconds === 'function';
// }

// export function isRRule(value: unknown): value is RRule {

// 	// @ts-ignore
// 	return value !== null && typeof value === 'object' && typeof value.between === 'function' && typeof value.toString === 'function';
// }

// export function toJSON(value: ICalDateTimeValue | null | undefined): string | null | undefined {
// 	if (!value) {
// 		return value;
// 	}
// 	if(typeof value === 'string') {
// 		return value;
// 	}

// 	return value.toJSON();
// }

/**
 * Converts a given duration in seconds to a iCalendar compatible durration
 * string.
 * @param seconds Duration in seconds
 * @returns Duration string
 */
export function toDurationString(seconds: number): string {
	let string = '';

	// < 0
	if (seconds < 0) {
		string = '-';
		seconds *= -1;
	}

	string += 'P';

	// DAYS
	if (seconds >= 86400) {
		string += Math.floor(seconds / 86400) + 'D';
		seconds %= 86400;
	}
	if (!seconds && string.length > 1) {
		return string;
	}

	string += 'T';

	// HOURS
	if (seconds >= 3600) {
		string += Math.floor(seconds / 3600) + 'H';
		seconds %= 3600;
	}

	// MINUTES
	if (seconds >= 60) {
		string += Math.floor(seconds / 60) + 'M';
		seconds %= 60;
	}

	// SECONDS
	if (seconds > 0) {
		string += seconds + 'S';
	} else if (string.length <= 2) {
		string += '0S';
	}

	return string;
}

export function filterEmpty<T>(unfiltered: Array<T>): Array<NonNullable<T>> {
	return unfiltered.filter(str => !!str) as Array<NonNullable<T>>;
}
