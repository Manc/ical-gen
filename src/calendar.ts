import { toDurationString } from './utils';
import ICalComponent from './component';

export interface ICalCalendarProdId {
	company: string;
	product: string;
	language: string;
}

export type ICalCalendarMethod =
	| 'PUBLISH'
	| 'REQUEST'
	| 'REPLY'
	| 'ADD'
	| 'CANCEL'
	| 'REFRESH'
	| 'COUNTER'
	| 'DECLINECOUNTER'
;

export type ICalCalendarOptions =
	Pick<ICalCalendar, 'prodId'> &
	Partial<
		Pick<ICalCalendar, 'method' | 'name' | 'description' | 'url' | 'ttl' | 'timezone' | 'children'>
	>;

export default class ICalCalendar extends ICalComponent {
	public tagName: string = 'VCALENDAR';

	/**
	 * The identifier for the product that created the iCalendar object.
	 * 
	 * This property is required. If a new `ICalCalendar` is instantiated
	 * without options, a default with `Unnamed` values will be applied.
	 */
	public prodId: ICalCalendarProdId;

	/**
	 * Method.
	 */
	public method?: ICalCalendarMethod;

	/**
	 * Name of the calendar.
	 */
	public name?: string;

	/**
	 * An optional description for the calendar. This is a non-standard property
	 * only supported by some clients.
	 * 
	 * The value will be used for the property `X-WR-CALDESC`.
	 */
	public description?: string;

	/**
	 * If the calendar is being used as a feed, its URL.
	 * @example http://example.com/my/feed.ical
	 */
	public url?: string;

	/**
	 * If the calendar is being used as a feed, its TTL in seconds, in order
	 * to give the calendar client application a hint how often to fetch new
	 * data from the server. For example, `86400` would suggest to update the
	 * feed daily.
	 * 
	 * The value will be used for the properties `REFRESH-INTERVAL` and
	 * `X-PUBLISHED-TTL`.
	 */
	public ttl?: number;

	/**
	 * Time zone for the calendar as a whole. Usually you should define the
	 * time zone for individual events of the calendar.
	 */
	public timezone?: string;

	constructor(options?: ICalCalendarOptions) {
		super(options);

		const {
			prodId,
			method,
			name,
			description,
			url,
			ttl,
			timezone,
		} = options || {
			prodId: {
				company: 'Unnamed',
				product: 'Unnamed',
				language: 'EN',
			}
		};

		this.prodId = prodId;
		this.name = name;
		this.method = method;
		this.description = description;
		this.url = url;
		this.ttl = ttl;
		this.timezone = timezone;
	}

	public getTimeZones = (): string[] => {
		const timeZones = new Set<string>(
			this.timezone ? [this.timezone] : undefined
		);

		(this.children.filter(child => child instanceof ICalComponent) as ICalComponent[])
			.forEach(child => {
				child.getTimeZones().forEach(eventZone => timeZones.add(eventZone));
			});

		return [...timeZones];
	};

	/**
	 * Render calendar instance as lines in iCalendar format.
	 */
	// private renderRawLines(): Array<string | undefined> {
	protected async renderToLines(): Promise<Array<ICalComponent | string | undefined>> {
		const {
			prodId,
			url,
			method,
			name,
			description,
			timezone,
			ttl,
			// events,
			// customLines,
			// children,
			getChildrenForRender,
			// vTimeZoneGenerator,
		} = this;
		const { company, product, language } = prodId;

		const ttlDuration = ttl && toDurationString(ttl);

		const lines: Array<ICalComponent | string | undefined> = [
			`PRODID:-//${company}//${product}//${language}`,
			'VERSION:2.0',
			'CALSCALE:GREGORIAN',
			url && `URL:${url}`,
			method && `METHOD:${method}`,
			description && `X-WR-CALDESC:${description}`,
			...(name ? [
				`NAME:${name}`,
				`X-WR-CALNAME:${name}`,
			] : []),
			...(ttlDuration ? [
				`REFRESH-INTERVAL;VALUE=DURATION:${ttlDuration}`,
				`X-PUBLISHED-TTL:${ttlDuration}`,
			] : []),
			...(timezone ? [
				`TIMEZONE-ID:${timezone}`,
				`X-WR-TIMEZONE:${timezone}`,
			] : []),
			...(await getChildrenForRender()),
		];

		return lines;
	}
}
