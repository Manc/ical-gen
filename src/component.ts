import { foldLine, filterAndJoinICalLines } from './utils';

export interface ICalComponentOptions {
	children?: Array<ICalComponent | string>;
};

export default abstract class ICalComponent {
	/**
	 * The `tagName` will result in opening and closing tag lines like this
	 * when the component is rendered:
	 * 
	 * ```
	 * BEGIN:tagName
	 * ...
	 * END:tagName
	 * ```
	 */
	public abstract readonly tagName: string;

	/**
	 * Child components.
	 */
	public children: Array<ICalComponent | string>;

	/**
	 * Default component constructor.
	 */
	constructor(options?: ICalComponentOptions) {
		this.children = options?.children || [];
	}

	public addChild(child: ICalComponent | string): this {
		this.children.push(child);
		return this;
	}

	/**
	 * This method must be implemented to return all time zone strings used
	 * by the instance if the component deals with time zones (e.g. an event
	 * component). If the component has no capabilities of introducing time
	 * zones, this method should return an empty array.
	 */
	public abstract getTimeZones: () => Array<string>;

	/**
	 * An opportunity for a component to dynamically manipulate the child
	 * components when the component is rendered. For example, it can be used
	 * to return more generated components based on the `children` property,
	 * or to filter components.
	 */
	public getChildrenForRender = (): Promise<Readonly<Array<ICalComponent | string>>> => {
		return Promise.resolve(this.children);
	};

	/**
	 * Render component instance as lines in iCalendar format. This method must
	 * be implmented by each component. It is called when the component is
	 * rendered.
	 */
	protected abstract renderToLines(): Promise<Array<ICalComponent | string | undefined>>;

	 /**
	 * Render the component and its child components to a string.
	 */
	public async renderToString(): Promise<string> {
		const unfilteredLines = await Promise.all(
			(await this.renderToLines()).map<Promise<string | undefined>>(async (row) => {
				if (typeof row === 'string') {
					return foldLine(row.trimEnd());
					// if (trimmed) {
					// 	return trimmed;
					// }
				} else if (row instanceof ICalComponent) {
					// const childLines = await row.renderToLines();
					return row.renderToString();
					// if (childStr) {
					// 	return childStr;
					// }
				}
			})
		);

		return filterAndJoinICalLines([
			foldLine(`BEGIN:${this.tagName}`),
			...unfilteredLines,
			foldLine(`END:${this.tagName}`),
		]);
	}
}
