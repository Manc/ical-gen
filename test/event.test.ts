// Import from `index.ts` to check that the main entry exports everything.
import { ICalEvent, ICalEventOptions } from '../src';
import { DateTime } from 'luxon';


describe('ICalEvent', () => {
	it('renders the provided options properly', async () => {
		const options: ICalEventOptions = {
			start: {
				date: new Date('2021-05-21T10:00:00Z'),
				zone: 'America/Los_Angeles',
			},
			end: {
				date: new Date('2021-05-21T11:30:00Z'),
				zone: 'America/Los_Angeles',
			},
			uid: 'testuid',
			sequence: 99,
			stamp: new Date('2021-05-09T07:08:09Z'),
			summary: 'Test Title',
			description: 'Test description.',
			url: 'https://example.com',
		};
		const str = await new ICalEvent(options).renderToString();
		expect(str).toMatchSnapshot();
	});

	it('handles event dates being native JS Dates with time zone strings', async () => {
		const options: ICalEventOptions = {
			start: {
				date: new Date('2021-05-21T10:00:06Z'),
				zone: 'America/Los_Angeles',
			},
			end: {
				date: new Date('2021-05-21T11:30:58Z'),
				zone: 'America/Los_Angeles',
			},
			sequence: 1,
			stamp: new Date(1621036800000),
			summary: 'Test',
			uid: 'testuid',
		};
		const str = await new ICalEvent(options).renderToString();
		expect(str).toContain('DTSTART;TZID=America/Los_Angeles:20210521T030006');
		expect(str).toContain('DTEND;TZID=America/Los_Angeles:20210521T043058');
	});

	it('handles event date being Luxon DateTime objects with time zones set for event with time', async () => {
		const options: ICalEventOptions = {
			start: DateTime.fromMillis(new Date('2021-05-21T10:00:06Z').getTime(), { zone: 'America/Los_Angeles' }),
			end: DateTime.fromMillis(new Date('2021-05-21T11:30:58Z').getTime(), { zone: 'America/Los_Angeles' }),
			sequence: 1,
			stamp: new Date(1621036800000),
			summary: 'Test',
			uid: 'testuid',
		};
		const str = await new ICalEvent(options).renderToString();
		expect(str).toContain('DTSTART;TZID=America/Los_Angeles:20210521T030006');
		expect(str).toContain('DTEND;TZID=America/Los_Angeles:20210521T043058');
	});

	it('handles event dates being native JS Dates for all-day event', async () => {
		const options: ICalEventOptions = {
			start: {
				date: new Date('2021-05-21T00:00:00Z'),
			},
			end: {
				date: new Date('2021-05-22T00:00:00Z'),
			},
			allDay: true,
			sequence: 1,
			stamp: new Date(1621036800000),
			summary: 'Test',
			uid: 'testuid',
		};
		const str = await new ICalEvent(options).renderToString();
		expect(str).toContain('DTSTART;VALUE=DATE:20210521');
		expect(str).toContain('DTEND;VALUE=DATE:20210522');
	});

	it('handles event dates being Luxon DateTime objects for all-day event', async () => {
		const options: ICalEventOptions = {
			start: DateTime.fromMillis(new Date('2021-05-21T00:00:00Z').getTime(), { zone: 'UTC' }),
			end: DateTime.fromMillis(new Date('2021-05-22T00:00:00Z').getTime(), { zone: 'UTC' }),
			allDay: true,
			sequence: 1,
			stamp: new Date(1621036800000),
			summary: 'Test',
			uid: 'testuid',
		};
		const str = await new ICalEvent(options).renderToString();
		expect(str).toContain('DTSTART;VALUE=DATE:20210521');
		expect(str).toContain('DTEND;VALUE=DATE:20210522');
	});
});
