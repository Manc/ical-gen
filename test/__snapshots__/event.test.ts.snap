// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ICalEvent renders the provided options properly 1`] = `
"BEGIN:VEVENT
UID:testuid
SEQUENCE:99
DTSTAMP:20210509T070809Z
DTSTART;TZID=America/Los_Angeles:20210521T030000
DTEND;TZID=America/Los_Angeles:20210521T043000
SUMMARY:Test Title
DESCRIPTION:Test description.
URL;VALUE=URI:https://example.com
END:VEVENT"
`;
